Remove in-app marketplace Search results tab and unify results into existing tabs

bor0 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

- Remove `<SearchResults />` component, remove it from the tabs, remove any references in the code
- Do the necessary changes so that search persists (when we change tabs), and is aligned with the current tab selection (addresses [this](https://github.com/woocommerce/woocommerce/pull/51266#pullrequestreview-2296043558)).
- Refactor fetching logic. On every search, we fetch all categories to keep the counts updated, and then do the category filtering in the frontend to display the results in the current tab.
    - This change allows us to show counts not only on `!! query.term` but also on `isLoading`, which addresses [this](https://github.com/woocommerce/woocommerce/pull/51266#pullrequestreview-2296429591)
- Apply CSS color fixes (addresses [this](https://github.com/woocommerce/woocommerce/pull/51266#pullrequestreview-2296429591))

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/Automattic/woocommerce.com/issues/21576.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

1. Go to http://localhost:8080/wp-admin/admin.php?page=wc-admin&path=%2Fextensions
2. Ensure no counts in the tabs are shown (because we are not searching)
3. Click on Extensions, then Themes, then Business services, then My subscriptions.
    - Make sure data is populated.
4. Search for "Bookings" while on My subscriptions tab.
    - Make sure that Extensions and Themes show 62 and 5 respectively (both have a light gray background with gray text)
    - Business services tab should disappear
5. Perform a hard reload.
    - State should be maintained and you should get same results.
6. Navigate to Extensions.
    - Background should turn to black, text to white. Data should be populated with the search.
7. Perform a hard reload.
    - State should be maintained and you should get same results.
8. Navigate to Themes and then search for "WooPayments"
    - Should get "Try searching again using a different term, or take a look at our recommendations below."
    - Business services tab should re-appear
9. Perform a hard reload.
    - State should be maintained and you should get same results.
10. Search for "Glowess" while on the Themes tab.
     - Extensions and Themes count should show as 0 and 1 respectively.
     - Business services should disappear
     - Data should be properly populated.
11. Navigate to Extensions
    - Should get "Try searching again using a different term, or take a look at our recommendations below."
12. Search for "LegalVision".
    - Business services should appear now, and counts should show 0, 0, 1 respectively.
13. Navigate to Business Services.
    - Should only get LegalVision.
14. Navigate to Extensions. Click on the Search text, empty it, and press enter.
    - Result count should disappear and you should get all extensions listed.
15. Get creative and try some more steps :) e.g. try a combination of filtering with a category

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

bor0 committed:
Remove search results component and any references to it

bor0 committed:
Persist current tab for searching, or default to extensions if tab is not set

bor0 committed:
Persist term when switching across tabs

bor0 committed:
Lint

bor0 committed:
When a search is initiated, fetch all categories to keep the tab counts up to date.

The necessary filtering to display data to the current screen will be performed on the frontend.

bor0 committed:
Apply correct colors to the tabs, as per design

bor0 committed:
Beyond query.term, also rely on isLoading so that search result counts don't jump

bor0 committed:
Address an issue when the user searches for something that returns no results in the business services tab

bor0 committed:
Changelog

bor0 committed:
Addressed :)

bor0 commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -40,169 +42,157 @@ export default function Content(): JSX.Element {
-			'extensions',
-			'themes',
-			'business-services',
+		const categories: Array< {
+			key: keyof SearchResultsCountType;
+			type: ProductType;
+		} > = [
+			{ key: 'extensions', type: ProductType.extension },
+			{ key: 'themes', type: ProductType.theme },
+			{ key: 'business-services', type: ProductType.businessService },
 		];
```
This can probably go in some more generic file. We're making this connection in several places.

bor0 committed:
Change key to category

bor0 commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -40,169 +42,157 @@ export default function Content(): JSX.Element {
+			{ category: 'extensions', type: ProductType.extension },
+			{ category: 'themes', type: ProductType.theme },
+			{ category: 'business-services', type: ProductType.businessService },
 		];
```
This can probably go in some more generic file. We're making this connection in several places.

bor0 commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -40,169 +42,157 @@ export default function Content(): JSX.Element {
+	}, [ query.term, query.category ] ); // Depend on term and category
+
+	// Filter the products based on the selected tab
+	useEffect( () => {
+		let filtered: Product[] | null;
+		switch ( selectedTab ) {
+			case 'extensions':
+				filtered = allProducts.filter(
+					( p ) => p.type === ProductType.extension
+				);
+				break;
+			case 'themes':
+				filtered = allProducts.filter(
+					( p ) => p.type === ProductType.theme
+				);
+				break;
+			case 'business-services':
+				filtered = allProducts.filter(
+					( p ) => p.type === ProductType.businessService
+				);
+				break;
+			default:
+				filtered = [];
+		}
+		setFilteredProducts( filtered );
+	}, [ selectedTab, allProducts ] );
```
If [`categories` was reusable enough](https://github.com/woocommerce/woocommerce/pull/51297/files#r1755107249), maybe we can be smarter here?

KokkieH commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -40,169 +42,157 @@ export default function Content(): JSX.Element {
-		setIsLoading,
-		query?.section,
-	] );
+	}, [ query.term, query.category ] ); // Depend on term and category
```
I'm getting a missing dependencies error for this line: 'setHasBusinessServices', 'setIsLoading', and 'setSearchResultsCount'

bor0 replied:
These are already present [here](https://github.com/woocommerce/woocommerce/blob/9c18ebd77fc46491e3bbc53400fc86900ce3bc39/plugins/woocommerce-admin/client/marketplace/components/content/content.tsx#L38-L40)

KokkieH replied:
They're used within the useEffect hook, but not included in the dependency array for that hook.

bor0 replied:
Hmm, what gives the "missing dependencies" error? Is it a common practice to include functions (that do not change) in the effect's dependency array?

KokkieH replied:
I'm getting the error in my IDE:

![content_tsx_—_woocommerce](https://github.com/user-attachments/assets/4b70a6d6-6a60-4848-9eb6-4de6b7a632ee)

Here's the [React docs](https://react.dev/reference/react/useEffect#specifying-reactive-dependencies) on the topic.

> Is it a common practice to include functions (that do not change)

The function doesn't change, but the value it returns can change, if my understanding is correct. Cursor suggests memoizing the functions with a useCallback hook if you don't want to trigger unnecessary re-renders.

bor0 replied:
Gotcha! Thanks. Fixed in 0b2d2dca6de5087b3b0ae5599f8bb0114cebc4ab

KokkieH commented on plugins/woocommerce-admin/client/marketplace/components/products/products.tsx:
```diff
@@ -167,11 +167,6 @@ export default function Products( props: ProductsProps ) {
 
 	return (
 		<div className={ containerClassName }>
-			{ selectedTab === 'search' && (
```
We should also remove the `selectedTab` import from the marketplace context, as well as the `title` and `productListTitleClassName` consts, as this is the only place they were being used.

bor0 replied:
Nice! There were a bunch of defined but unused stuff and I did some neat cleanup here in a10eb02a40

KokkieH reviewed (commented):
Category filtering appears to be completely broken. With or without a search term, the product list doesn't update, and the content also jumps while switching categories.

https://github.com/user-attachments/assets/1210bb5a-9410-4601-babd-20b54d03b91b

> Business services tab should disappear

We originally hid the tab on no results because we didn't have any business services at the time we updated in-app to support them, but now that the product type is launched, I wonder if we still need it. Not hiding the tab will also prevent the current issue where if you do a refresh while on the Business Services tab, you get kicked back to the Extensions tab.

However, currently there aren't any featured business services set up to display in the no-results state, though, so it might make sense to keep it hidden until we have more overall.

Showing no results while the search is running isn't great, but that will be fixed in https://github.com/Automattic/woocommerce.com/issues/21596

Holding off on a full code review so the category filtering can be fixed first, but mentioning a couple small issues my IDE highlighted in files I happened to already have open when checking out this branch.

bor0 committed:
Fix category filter being broken

Whenever a category is requested, we need to do an additional request with the category param being the current category (overriding extensions/theme/business services).

Ideally the backend API would make a distinction between type (theme/extension/business service) and category, but this hack should do for now.

bor0 committed:
Lint

bor0 committed:
Remove unused variables h/t @KokkieH

bor0 commented:
Hi @KokkieH, thanks for taking the time to look!

> Category filtering appears to be completely broken.

I missed this special case of handling. I remember testing it and working with a prior commit, but now it should be good after 05fb7f1be355d71d7aeaf578f626be1b10d753b5 (ignore whitespace can help to better see the changes)

> the content also jumps while switching categories.

I noticed this happening on `trunk` as well. Maybe we can tackle a separate issue since this one is already becoming enormous?

KokkieH commented:
Found an interesting edge-case just clicking around: 
![Extensions_‹_WooCommerce_‹_WooCommerce_Core_E2E_Test_Suite](https://github.com/user-attachments/assets/5d8b6bff-1efe-462a-82c7-834b115965d4)

On the Extensions tab, search for "payments". Click through the category filters, until you reach a category that doesn't have any results for the current term. Get stuck with the no results state until you switch tabs.

This is happening because we still show ALL the categories returned by the original query to the `/categories` endpoint, even if there are no results for a given term in a category.

I think we either need to hide a category filter if it has no products (I have a feeling that will be rather complicated to achieve), or else ensure the category filter is still visible even in the no-results state.

bor0 committed:
Lint

bor0 committed:
Revert "Lint"

This reverts commit 0b2d2dca6de5087b3b0ae5599f8bb0114cebc4ab.

bor0 committed:
Actually fix lint without introducing infinite loop

Reproducible at http://localhost:8080/wp-admin/admin.php?page=wc-admin&term=payments&tab=extensions&path=%2Fextensions&category=customer-service

bor0 committed:
Show category selector even if there are no results displayed

bor0 commented:
> On the Extensions tab, search for "payments". Click through the category filters, until you reach a category that doesn't have any results for the current term. Get stuck with the no results state until you switch tabs.

I noticed this happening on `trunk` as well and had assumed it was the expected behavior.

> I have a feeling that will be rather complicated to achieve

Yeah, we are hitting the endpoint only after a category was chosen. The fix for the other approach is straightforward and I applied that in fe1b6ece97.

KokkieH commented:
> had assumed it was the expected behavior

I mean, it feels a bit unexpected 😆 

Thanks for the quick fix.

KokkieH commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -40,169 +42,202 @@ export default function Content(): JSX.Element {
+					setIsLoading( false );
+				} );
+		} else {
+			// Fetch all categories when query.term or query.category changes
```
Just to clarify, by "all categories" here you mean all three product types, i.e. extensions, themes, and business services? It seems so, going by the requests I see in the network tab when doing a search.

In that case I recommend updating the comment to make that clearer. "All categories" can be interpreted to refer to the actual categories appearing in the filters on each tab.

raicem replied:
I got confused by the same thing. Is it possible to rename this as `tabs`? I know this list does not cover all tabs, but `extensions`, `themes`, and `business-services` are described as tabs in other files

bor0 replied:
Nice catch! Addressed in 205ab7870a

KokkieH commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -40,169 +42,202 @@ export default function Content(): JSX.Element {
+			)
+				.then( ( results ) => {
+					const combinedProducts = results.flat();
+					setAllProducts( combinedProducts );
```
So we have one state variable storing results from all three queries. And in the next useEffect we just filter that list based on the selected tab?

bor0 replied:
That's correct!

KokkieH reviewed (approved):
When using the category filters on a given tab, the counts for the other tabs go to 0 (because obviously extensions, themes and services don't use the same categories). If it's not too complex, it might be good if category filtering didn't affect the counts for other tabs, but it's not a blocker.

When using a screenreader, if I search for something, there is no announcement once the search completes that anything has happened. It would be good if we could announce something like "62 Extensions found", i.e. announcing the current tab, and results for that tab.

This should definitely be fixed, but given the size of this PR we can do that in a follow-up task.

Otherwise everything appears to work, and I don't spot any issues with the code, though a second review would be good in case I missed anything.

raicem commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -40,169 +42,202 @@ export default function Content(): JSX.Element {
+				.finally( () => {
+					setIsLoading( false );
+				} );
+		} else {
```
I'm not a massive fan of big `else` statements. They might be difficult to follow and this else block has a lot going on. 

As a first step, would it make sense to create a new `fetchSearchResultsForAllCategories` helper function that will include the `Promise.all` code to reduce the else block? Just as a brainstorming exercise, no need to implement it right away

bor0 replied:
This is a really good catch! I also didn't like the blocks within this conditional. As testing steps are complex and since this PR is blocking several other PRs, I will merge as is. Created https://github.com/Automattic/woocommerce.com/issues/21700 to address this and other refactor suggestions.

raicem reviewed (commented):
Thank you for the PR @bor0, a big change handled gracefully. 

> Search for "Bookings" while on My subscriptions tab. 

After this, I was expecting it get the Extensions tab like it's described in the issue, but I didn't. When I make a search in the Discover tab, I do get the Extensions tab. So maybe a bug with My Subscriptions page?

bor0 committed:
Update comment to be less misleading

bor0 committed:
Query isn't used here
